File: spark/src/main/scala/geotrellis/spark/io/avro/AvroRecordCodec.scala
package geotrellis.spark.io.avro

import org.apache.avro.generic._
import org.apache.avro._
import scala.annotation.implicitNotFound
import scala.reflect.ClassTag

@implicitNotFound("Cannot find AvroRecordCodec for ${T}. Try to import geotrellis.spark.io.avro.codecs._")
abstract class AvroRecordCodec[T: ClassTag] extends AvroCodec[T, GenericRecord] {
  def schema: Schema
  def encode(thing: T, rec: GenericRecord)
  def decode(rec: GenericRecord): T

  def encode(thing: T): GenericRecord = {
    val rec = new GenericData.Record(schema)
    encode(thing, rec)
    rec
  }

  def supported[O](other: O): Boolean = {
    implicitly[ClassTag[T]].unapply(other).isDefined
  }
}

File: geotrellis-python/geotrellis/spark/io/avro/AvroRecordCodec.py
from geotrellis.spark.io.avro.AvroCodec import AvroCodec

class AvroRecordCodec(AvroCodec):
    #def __init__(self, record_type):
    #    AvroCodec.__init__(self, record_type):
    @property
    def schema(self):
        pass
    def _encode(self, thing, dct):
        pass
    def decode(self, dct):
        pass
    def encode(self, thing, dct = None):
        if dct is None:
            dct = {}
        self._encode(thing, dct)
        return dct
